server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -0,0 +1,967 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing.allocation.allocator;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.IntroSorter;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.routing.RoutingNode;
+import org.opensearch.cluster.routing.RoutingNodes;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.ShardRoutingState;
+import org.opensearch.cluster.routing.UnassignedInfo;
+import org.opensearch.cluster.routing.allocation.AllocateUnassignedDecision;
+import org.opensearch.cluster.routing.allocation.AllocationDecision;
+import org.opensearch.cluster.routing.allocation.MoveDecision;
+import org.opensearch.cluster.routing.allocation.NodeAllocationResult;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.cluster.routing.allocation.decider.AllocationDeciders;
+import org.opensearch.cluster.routing.allocation.decider.Decision;
+import org.opensearch.cluster.routing.allocation.decider.DiskThresholdDecider;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.gateway.PriorityComparator;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.StreamSupport;
+
+import static org.opensearch.cluster.routing.ShardRoutingState.RELOCATING;
+
+/**
+ * A {@link LocalShardsBalancer} used by the {@link BalancedShardsAllocator} to perform allocation operations
+ * for local shards within the cluster.
+ *
+ * @opensearch.internal
+ */
+public class LocalShardsBalancer extends ShardsBalancer {
```

Comment:
I'm not sure how conservative we need to be here, but in the interest of compatibility it seems like we could leave the inner class 'Balancer' in place with its current name and just change it to extend the newly introduced interface. These things are tagged `opensearch.internal` but that doesn't mean they aren't used somewhere (though I couldn't find any usage of Balancer in any other repo in opensearch-project). What do you think?

Reply:
1. Why do we want to expose the `Balancer` outside `BalancedShardAllocator`
2. Can we use a factory for `Balancer` creation?

Reply:
> Why do we want to expose the Balancer outside BalancedShardAllocator

I don't think we want to, but it is already used in `org.opensearch.cluster.routing.allocation.AllocationConstraints` which requires it to be public.

> Can we use a factory for Balancer creation?

The approach here (#4759) calls for having 2 balancers with one running after the other, so I'm not sure a factory would help.

Reply:
> I'm not sure how conservative we need to be here, but in the interest of compatibility it seems like we could leave the inner class 'Balancer' in place with its current name and just change it to extend the newly introduced interface. These things are tagged `opensearch.internal` but that doesn't mean they aren't used somewhere (though I couldn't find any usage of Balancer in any other repo in opensearch-project). What do you think?

I wanted to improve the readability and maintainability by splitting it across appropriate classes. We can keep `LocalShardsBalancer` and `BalancedShardsAllocator.Balancer` and deprecate the latter. We can get rid of it with 3.0.0.